homeassistant/components/tplink/__init__.py

```diff
@@ -0,0 +1,147 @@
+            DOMAIN, context={'source': config_entries.SOURCE_IMPORT}))
+
+        return True
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up TPLink from a config entry."""
+    from pyHS100 import SmartBulb, SmartPlug, SmartDeviceException
+
+    devices = {}
+
+    config_data = hass.data[DOMAIN][ATTR_CONFIG]
+
+    # These will contain the initialized devices
+    lights = hass.data[DOMAIN][CONF_LIGHT] = []
+    switches = hass.data[DOMAIN][CONF_SWITCH] = []
+
+    # If discovery is defined and not disabled, discover devices
+    # If initialized from configure integrations, there's no config
+    # so we default here to True
+    if config_data is None or config_data[CONF_DISCOVERY]:
+        devs = await _async_has_devices(hass)
+        _LOGGER.info("Discovered %s TP-Link smart home device(s)", len(devs))
+        devices.update(devs)
+
+    def _device_for_type(host, type_):
+        dev = None
+        if type_ == CONF_LIGHT:
+            dev = SmartBulb(host)
+        elif type_ == CONF_SWITCH:
+            dev = SmartPlug(host)
+
+        return dev
+
+    # When arriving from configure integrations, we have no config data.
+    if config_data is not None:
+        for type_ in [CONF_LIGHT, CONF_SWITCH]:
+            for entry in config_data[type_]:
+                try:
+                    host = entry['host']
+                    dev = _device_for_type(host, type_)
+                    devices[host] = dev
+                    _LOGGER.debug("Succesfully added %s %s: %s",
+                                  type_, host, dev)
+                except SmartDeviceException as ex:
+                    _LOGGER.error("Unable to initialize %s %s: %s",
+                                  type_, host, ex)
+
+    # This is necessary to avoid I/O blocking on is_dimmable,
+    # the upstream library
+    def _fill_device_lists():
+        for dev in devices.values():
+            if isinstance(dev, SmartPlug):
+                if dev.is_dimmable:  # Dimmers act as lights
+                    lights.append(dev)
+                else:
+                    switches.append(dev)
+            elif isinstance(dev, SmartBulb):
+                lights.append(dev)
+            else:
+                _LOGGER.error("Unknown smart device type: %s", type(dev))
+
+    # Avoid blocking on is_dimmable
+    await hass.async_add_executor_job(_fill_device_lists)
+
+    forward_setup = hass.config_entries.async_forward_entry_setup
+    if lights:
+        hass.async_create_task(forward_setup(config_entry, 'light'))
+    if switches:
+        hass.async_create_task(forward_setup(config_entry, 'switch'))
+
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    forward_unload = hass.config_entries.async_forward_entry_unload
+    remove_lights = await forward_unload(entry, 'light')
+    remove_switches = await forward_unload(entry, 'switch')
+
+    hass.data[DOMAIN][CONF_LIGHT] = None
```

Comment:
We can just pop or delete the domain key from hass.data.

Reply:
Changed.